### PR TITLE
Request, Apply & Claim Withdraw

### PR DESCRIPTION
**Pull request by bh2smith**

### CODE REVIEW

File: contracts/SnappBase.sol [hunk 1]
```diff
@@ -117,6 +147,72 @@ contract SnappBase is Ownable {
...
+        emit StateTransition(TransitionType.Deposit, this.stateIndex(), _newStateRoot, slot);
+    }
+
+    function requestWithdrawal(uint8 tokenId, uint amount) public onlyRegistered() {
+        require(amount != 0, "Must request positive amount");
+
+        address tokenAddress = tokenIdToAddressMap[tokenId];
+        require(tokenAddress != address(0), "Requested token is not registered");
+        require(
+            ERC20(tokenAddress).balanceOf(address(this)) >= amount,
```
Comment (josojo) [inline contracts/SnappBase.sol:179]: I don't know whether these kinds of sanity checks make sense. In most cases, they will just cost gas.

Comment (josojo) [reply contracts/SnappBase.sol:179]: But we would somehow need to check that the amount is not generating an overflow. In the snark, we will only be able to deal with a certain precision.

**Review by josojo — CHANGES_REQUESTED** [review]
Great start!

Like the code.

I would do one thing differently: not sure about it.
I would require that all the applyWithdrawals are handed over to the smart contract in sequence. E.g. if the applyWithdrawal is not handed in for slot 4, we do not accept an applyWithdrawal for slot 5. 
This makes just everything easier to reason about and we have the same structure as with deposits.

If a successful challenge happens of a withdrawal in slot x, we would then only have to kill all withdrawals in the array from x and reset the state to the current state root.

If a successful challenge happens on any other state transition (with slot y),  then we will delete all withdrawal requests with a higher appliedAccountStateIndex than y.

The sequential applyWithdrawals enforcement has the benefit that the claimableWithdraws array is ordered by stateIndex and hence all operations are easier to perform.

Please also add the code for an actual claim in this pr to see the whole system at once.

**Comment by josojo** [comment]
Ups... just checked the current contract code. Actually, the applyDeposits are not forced to be in sequence. Maybe we should do it for an easier roll-back. 

I am wondering whether we should enforce it. Have you thought about rollbacks of deposits?

File: contracts/SnappBase.sol (hunk 1 above)
Comment (bh2smith) [reply contracts/SnappBase.sol:179]: What do you mean by overflow here? There is a limit to the number of withdraws per slot?

**Comment by bh2smith** [comment]
Yes, I can certainly impose that deposits and withdraws are applied in order. I will start with the Withdraws, because there is already enough information contained in the data structure to implement this easily. I think it will be better to impose this on deposits in a separate PR. I will do that immediately after I am finished with this one. This will involve using a similar structure for depositStates as for pendingWithdraws (as described in Felix's Gist).

File: contracts/SnappBase.sol (hunk 1 above)
Comment (josojo) [reply contracts/SnappBase.sol:179]: no..., the variable amount will be represented by some amount of bits, maybe 100, but the uint256 has 256 bits. So there is the possibility to overflow the snark 100 bits

Comment (bh2smith) [reply contracts/SnappBase.sol:179]: Do we have any estimate on what the snark can handle?

File: test/snapp_base.js [hunk 2]
```diff
@@ -116,6 +116,8 @@ contract("SnappBase", async (accounts) => {
 
     it("Can't exceed max tokens", async () => {
       const instance = await SnappBase.new()
+
+      SnappBase.ca
```
Comment (josojo) [inline test/snapp_base.js]: what is this? I have seen it several times :)

**Comment by josojo** [comment]
Looks really really good

I would only change some thing: we should not allow immediate claims. We want that our chain first "finalizes" and no roll-back can mess it up.

Maybe we should just start by inserting something like this
require( claimableWithdraws[slot].appliedStateIndex +300 <block)

File: test/snapp_base.js (hunk 2 above)
Comment (bh2smith) [reply test/snapp_base.js]: So weird. I don't see it anywhere...

Comment (bh2smith) [reply test/snapp_base.js]: Found it and removed it, please let me know if you see it anywhere else.

**Review by josojo — APPROVED** [review]
Solid work!

**Comment by fleupold** [comment]
This PR is > 1000 lines and has a few different things going on (e.g. adding unit tests for view functions) which makes it very hard to review.
It would have been nice to split e.g. importing the Merkle library + unit tests into a separate PR and also move changed unrelated to the withdraw functionality in a separate one.

**Comment by bh2smith** [comment]
@fleupold I agree. What should we do about it now?

**Comment by fleupold** [comment]
Nothing, just feedback for next time

**Pull request merged**